main.go

```diff
@@ -0,0 +1,260 @@
+	if err != nil {
+		log.Fatalf("could not write file %s: %s", *outputFile, err.Error())
+	}
+}
+
+func parseSources() interface{} {
+	output, sources := parseInput()
+	var repositories []repository
+	if output["repositories"] != nil && len(output["repositories"].([]interface{})) > 0 {
+		switch t := output["repositories"].(type) {
+		case []interface{}:
+			for _, v := range t {
+				var repo repository
+				err := mapstructure.Decode(v, &repo)
+				if err != nil {
+					log.Errorf("could not map `%+v` repository struct, skipping", v)
+					continue
+				}
+				repositories = append(repositories, repo)
+			}
+		}
+	}
+
+	for _, source := range sources {
+		var repos []repository
+		switch source.SourceType {
+		case "gitlab":
+			repos = parseGitlab(source)
+		case "github":
+			repos = parseGithub(source)
+		case "azdo":
+			repos = parseAzDO(source)
+		default:
+			log.Errorf("could not parse source %s with unknown type %s", source.SourceIdent, source.SourceType)
+			continue
+		}
+		repositories = append(repositories, repos...)
+	}
+	output["repositories"] = repositories
+	return output
+}
+
+func parseAzDO(s source) (repositories []repository) {
+	log.Debugf("downloading Azure DevOps projects for %s", s.SourceIdent)
+	idents := strings.Split(s.SourceIdent, "/")
+	org := strings.Join(idents[:len(idents)-1], "/")
+	project := idents[len(idents)-1]
+
+	connection := azuredevops.NewPatConnection(org, s.SourceAuth)
+
+	ctx := context.Background()
+
+	client, err := git.NewClient(ctx, connection)
+	if err != nil {
+		log.Fatal(err)
+	}
+	repos, err := client.GetRepositories(ctx, git.GetRepositoriesArgs{Project: &project})
+	if err != nil {
+		log.Errorf("could not fetch azdo repositories %s", err.Error())
+	}
+
+	for _, repo := range *repos {
+		//isDisabled attribute is available from API version 7.1 which is currently in preview
+		file, err := client.GetItem(ctx, git.GetItemArgs{
+			RepositoryId: gitlab.String(repo.Id.String()),
+			Path:         gitlab.String("/" + composerPath),
+		})
+		if file == nil || err != nil {
+			continue
+		}
+		repositories = append(
+			repositories,
+			repository{
+				Name: *repo.Name,
+				Type: "git",
+				URL:  *repo.SshUrl,
+			},
+		)
+	}
+
+	return repositories
+}
+
+func parseGithub(s source) (repositories []repository) {
+	log.Debugf("downloading github projects for organization %s", s.SourceIdent)
+	var tc *http.Client
+	ctx := context.Background()
+	if s.SourceAuth != "" {
+		ts := oauth2.StaticTokenSource(
+			&oauth2.Token{AccessToken: s.SourceAuth},
+		)
+		tc = oauth2.NewClient(ctx, ts)
+	}
+
+	client := github.NewClient(tc)
+
+	options := github.RepositoryListByOrgOptions{
+		ListOptions: github.ListOptions{
+			Page:    1,
+			PerPage: 100,
+		},
+	}
+	for {
+		repos, response, err := client.Repositories.ListByOrg(ctx, s.SourceIdent, &options)
+		if err != nil {
+			log.Errorf("could not fetch github response %s", err.Error())
+		}
+		for _, repo := range repos {
+			if repo.GetArchived() {
```

Comment:
A little too much immersion, it's kinda hard to read. I'd move it out in a separate function.

Reply:
valid point, thanks! I've extracted repo prepare and get client methods